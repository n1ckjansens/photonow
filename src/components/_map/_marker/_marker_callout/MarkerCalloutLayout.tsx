import React, { useRef, useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
	faDollarSign,
	faCamera,
	faArrowLeft,
	faArrowRight,
} from '@fortawesome/free-solid-svg-icons'

//SLIDER PROP ONLY FOR TESTING PURSPOSES IN NOT TESTING MODE IT SHOLD BE FALSY
const MarkerCalloutLayout = ({testSlider = null}) => {
	//useState is used to dynamicly select slider element when page loaded (initial state is a custom slider.scroll event FOR TESTING ONLY)
	const [slider, setSlider] = useState<HTMLDivElement | null>(testSlider)

	//linking slider element with reference
	const sliderRef = useRef<HTMLDivElement>(null)

	//Finding slider element in DOM when page loaded
	React.useEffect(() => {
		const slider = sliderRef.current
		setSlider(slider)
	}, [])

	//Function to scroll slider to the left when user pressed left arrow
	const leftArrowPressed = () => {
		slider && slider.scroll({ left: slider.scrollLeft - 130, behavior: 'smooth' })
	}

	//Function to scroll slider to the right when user pressed right arrow
	const rightArrowPressed = () => {
		slider && slider.scroll({ left: slider.scrollLeft + 130, behavior: 'smooth' })
	}

	return (
		<>
			<div className="user__credentials">
				<div className="user__avatar">
					<div className="user__type">
						<FontAwesomeIcon icon={faCamera} />
					</div>
					<img
						className="user__avatar__image"
						src="https://sun1-47.userapi.com/impg/c856124/v856124902/1fba84/39E3gZWT5qk.jpg?size=200x0&quality=90&crop=416,0,1334,1334&sign=8e096f2c133c95e3720669fa1b84b1f5&ava=1"
						alt="user__avatar"
					/>
				</div>
				<h2 className="user__name">Мария Погорелова</h2>
			</div>
			<div className="user__credentials__secondary">
				<p className="user__address">Москва, пр-т Вернадского 97 к2</p>
				<p className="user__work__hours">Ежедневно 9:00 - 21:00</p>
			</div>
			<div className="user__portfolio">
				<div className="user__portfolio__nav__buttons">
					<div
						className="nav__button"
						id="toggle__previous__button"
						onClick={leftArrowPressed}
					>
						<FontAwesomeIcon icon={faArrowLeft} />
					</div>
					<div
						className="nav__button"
						id="toggle__next__button"
						onClick={rightArrowPressed}
					>
						<FontAwesomeIcon icon={faArrowRight} />
					</div>
				</div>
				<div className="portfolio__items" ref={sliderRef}>
					<div className="portfolio__item">
						<img
							src="https://sun1-92.userapi.com/8_i2e2SRmAretsIzWFrOITKsCg1yCZTht9J6VQ/4gyVPk9RqkE.jpg"
							alt=""
						/>
					</div>
					<div className="portfolio__item">
						<img
							src="https://sun1-22.userapi.com/YL7rAS7qsINWZ2I9EbqAVQTzCmDVeuA0yt2ciQ/DGrGGwLcHqE.jpg"
							alt=""
						/>
					</div>
					<div className="portfolio__item">
						<img
							src="https://sun1-21.userapi.com/w4JtHiwQHKYxWeSP1lt5NXpp8B9xPXsPMzcvjw/2tvqFEHmkEk.jpg"
							alt=""
						/>
					</div>
					<div className="portfolio__item">
						<img
							src="https://sun1-97.userapi.com/dDKYel8QHlynDQCpQLjp5BXivNSWQKqdnUfPzA/B6MhQBga3tE.jpg"
							alt=""
						/>
					</div>
					<div className="portfolio__item">
						<img
							src="https://sun1-26.userapi.com/9Tr8qUTUVLXUUtev6S9olEQG06dgviog0EIP_g/TxvYsyy-GGY.jpg"
							alt=""
						/>
					</div>
				</div>
			</div>
			<div className="user__pricing">
				<FontAwesomeIcon className="icon" icon={faDollarSign} />
				<p className="user__pricing__alert">
					Если цена не указана, фотограф свяжется с вами и уточнит цену
				</p>
			</div>
			<div className="user__services">
				<b className="user__service">Семейная фотосессия</b>
				<b className="user__service">Интерьерная сьемка</b>
			</div>
			<div className="internal__service__features">
				<div className="feature">Безопасная сделка</div>
				<div className="feature">Подарки</div>
			</div>
			<button className="user__contact__button">Написать</button>
		</>
	)
}

export default MarkerCalloutLayout
