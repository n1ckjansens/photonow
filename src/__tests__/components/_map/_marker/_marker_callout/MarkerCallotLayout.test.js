import React from 'react'
import { shallow } from 'enzyme'
import MarkerCalloutLayout from '../../../../../components/_map/_marker/_marker_callout/MarkerCalloutLayout'

//Function to render component into test
const setUp = (props) => shallow(<MarkerCalloutLayout {...props} />)

describe('Marker callout popup component test', () => {
	let component,
		componentToTestButtons,
		buttonLeft,
		buttonRight,
		slider,
		fn,
		useEffect

	//Function to fake React.useEffect event
	const mockUseEffect = () => {
		useEffect.mockImplementationOnce((f) => f())
	}

	//Rendering component before each test
	beforeEach(() => {
		//Function to test onClick events working
		fn = jest.fn()
		//Rendering component to test ONLY proper rendering
		component = setUp()
		//This rendered component contains custom slider prop to test onClick events
		componentToTestButtons = setUp({ testSlider: { scroll: fn } })
		//Left slider button component
		buttonLeft = componentToTestButtons.find('#toggle__previous__button')
		//Right slider button component
		buttonRight = componentToTestButtons.find('#toggle__next__button')
		//Slider component
		slider = component.find('.portfolio__items')
		//Facking React.useEffect event
		useEffect = jest.spyOn(React, 'useEffect')
		mockUseEffect()
	})

	//Testing for successfull rendering of callout through .callout__container  wrapper
	it('Should render successfully and contain .user__credentials wrapper', () => {
		const user__credentials = component.find('.user__credentials')
		expect(user__credentials.length).toBe(1)
	})

	//Testing that callout contains slider component
	it('Should contain slider component', () => {
		expect(slider.length).toBe(1)
	})

	//Testing that onClick function called when pressing right button
	it('Slider right button should work', () => {
		buttonRight.simulate('click')
		expect(fn.mock.calls.length).toBe(1)
	})

	//Testing that onClick function called when pressing left button
	it('Slider left button should work', () => {
		buttonLeft.simulate('click')
		expect(fn.mock.calls.length).toBe(1)
	})

	//Testing component matches previos snapshots
	it('Should match snapshot', () => {
		expect(component).toMatchSnapshot()
	})
})
