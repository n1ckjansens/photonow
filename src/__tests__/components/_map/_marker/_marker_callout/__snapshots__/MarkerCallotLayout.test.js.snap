// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker callout popup component test Should match snapshot 1`] = `
<Fragment>
  <div
    className="user__credentials"
  >
    <div
      className="user__avatar"
    >
      <div
        className="user__type"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f030",
                "M512 144v288c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V144c0-26.5 21.5-48 48-48h88l12.3-32.9c7-18.7 24.9-31.1 44.9-31.1h125.5c20 0 37.9 12.4 44.9 31.1L376 96h88c26.5 0 48 21.5 48 48zM376 288c0-66.2-53.8-120-120-120s-120 53.8-120 120 53.8 120 120 120 120-53.8 120-120zm-32 0c0 48.5-39.5 88-88 88s-88-39.5-88-88 39.5-88 88-88 88 39.5 88 88z",
              ],
              "iconName": "camera",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </div>
      <img
        alt="user__avatar"
        className="user__avatar__image"
        src="https://sun1-47.userapi.com/impg/c856124/v856124902/1fba84/39E3gZWT5qk.jpg?size=200x0&quality=90&crop=416,0,1334,1334&sign=8e096f2c133c95e3720669fa1b84b1f5&ava=1"
      />
    </div>
    <h2
      className="user__name"
    >
      Мария Погорелова
    </h2>
  </div>
  <div
    className="user__credentials__secondary"
  >
    <p
      className="user__address"
    >
      Москва, пр-т Вернадского 97 к2
    </p>
    <p
      className="user__work__hours"
    >
      Ежедневно 9:00 - 21:00
    </p>
  </div>
  <div
    className="user__portfolio"
  >
    <div
      className="user__portfolio__nav__buttons"
    >
      <div
        className="nav__button"
        id="toggle__previous__button"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f060",
                "M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z",
              ],
              "iconName": "arrow-left",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </div>
      <div
        className="nav__button"
        id="toggle__next__button"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f061",
                "M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z",
              ],
              "iconName": "arrow-right",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </div>
    </div>
    <div
      className="portfolio__items"
    >
      <div
        className="portfolio__item"
      >
        <img
          alt=""
          src="https://sun1-92.userapi.com/8_i2e2SRmAretsIzWFrOITKsCg1yCZTht9J6VQ/4gyVPk9RqkE.jpg"
        />
      </div>
      <div
        className="portfolio__item"
      >
        <img
          alt=""
          src="https://sun1-22.userapi.com/YL7rAS7qsINWZ2I9EbqAVQTzCmDVeuA0yt2ciQ/DGrGGwLcHqE.jpg"
        />
      </div>
      <div
        className="portfolio__item"
      >
        <img
          alt=""
          src="https://sun1-21.userapi.com/w4JtHiwQHKYxWeSP1lt5NXpp8B9xPXsPMzcvjw/2tvqFEHmkEk.jpg"
        />
      </div>
      <div
        className="portfolio__item"
      >
        <img
          alt=""
          src="https://sun1-97.userapi.com/dDKYel8QHlynDQCpQLjp5BXivNSWQKqdnUfPzA/B6MhQBga3tE.jpg"
        />
      </div>
      <div
        className="portfolio__item"
      >
        <img
          alt=""
          src="https://sun1-26.userapi.com/9Tr8qUTUVLXUUtev6S9olEQG06dgviog0EIP_g/TxvYsyy-GGY.jpg"
        />
      </div>
    </div>
  </div>
  <div
    className="user__pricing"
  >
    <FontAwesomeIcon
      border={false}
      className="icon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            288,
            512,
            Array [],
            "f155",
            "M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z",
          ],
          "iconName": "dollar-sign",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <p
      className="user__pricing__alert"
    >
      Если цена не указана, фотограф свяжется с вами и уточнит цену
    </p>
  </div>
  <div
    className="user__services"
  >
    <b
      className="user__service"
    >
      Семейная фотосессия
    </b>
    <b
      className="user__service"
    >
      Интерьерная сьемка
    </b>
  </div>
  <div
    className="internal__service__features"
  >
    <div
      className="feature"
    >
      Безопасная сделка
    </div>
    <div
      className="feature"
    >
      Подарки
    </div>
  </div>
  <button
    className="user__contact__button"
  >
    Написать
  </button>
</Fragment>
`;
